### PR TITLE
Store trait associated items in fst

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Store imported traits' associated function/methods and constants into `ImportMap.fst` and pefrorm the imports search on them.

This is a first step towards trait autoimport during completion functionality, the way I see it, after this PR, only a few major things are left to be done:

* store all traits' assoc items into fst, not only the ones in scope, as we do now. Any code pointers on how to do this are welcome 😄 
* adjust a few modules in completions crate (`dot.rs`, `qualified_path.rs` at least) to query the import map, reusing the `import_assets` logic heavily

==
With the current import and autoimport implementations, it looks like for a single query, we're either interested in either associated items lookup or in all other `fst` contents lookup, but never both simultaneously.
I would rather not split `fst` in two but add another `Query` parameter to separate those, but let me know if you have any ideas.

### CODE REVIEW

File: crates/completion/src/completions/unqualified_path.rs [hunk 1]
```diff
@@ -143,6 +143,14 @@ fn fuzzy_completion(acc: &mut Completions, ctx: &CompletionContext) -> Option<()
...
         true,
     )
+    .filter(|import_candidate| match import_candidate {
+        Either::Left(ModuleDef::Function(function)) => function.as_assoc_item(ctx.db).is_none(),
+        Either::Left(ModuleDef::Const(const_)) => const_.as_assoc_item(ctx.db).is_none(),
+        Either::Left(ModuleDef::TypeAlias(type_alias)) => {
+            type_alias.as_assoc_item(ctx.db).is_none()
+        }
+        _ => true,
+    })
```
Comment (matklad) [inline crates/completion/src/completions/unqualified_path.rs:153]: Let's push this check down -- from within ImportMap, we can figure this out without extra queries. 

So, lets add an extra param to `find_similar_import`. It's even time to refactor it to CommandObject pattern, a-la FindUsage.

File: crates/hir_def/src/import_map.rs [hunk 2]
```diff
@@ -176,10 +178,31 @@ impl ImportMap {
...
+                // qualifier, ergo no need to store it for imports in import_map
+                AssocItemId::TypeAliasId(_) => continue,
+            };
+            let assoc_item = if is_type_in_ns {
+                ItemInNs::Types(module_def_id)
+            } else {
+                ItemInNs::Values(module_def_id)
+            };
+
+            let mut assoc_item_info = original_import_info.to_owned();
```
Comment (matklad) [inline crates/hir_def/src/import_map.rs:202]:
```suggestion
            let mut assoc_item_info = original_import_info.clone();
```

File: crates/hir_def/src/import_map.rs [hunk 3]
```diff
@@ -304,11 +327,11 @@ impl Query {
     }
 }
 
-fn contains_query(query: &Query, input_path: &ImportPath, enforce_lowercase: bool) -> bool {
-    let mut input = if query.name_only {
-        input_path.segments.last().unwrap().to_string()
+fn import_matches_query(import: &ImportInfo, query: &Query, enforce_lowercase: bool) -> bool {
```
Comment (matklad) [inline crates/hir_def/src/import_map.rs:330]: could be a method on Query

**Comment by matklad** [comment]
>store all traits' assoc items into fst, not only the ones in scope, as we do now. Any code pointers on how to do this are welcome smile

Hm, I don't quite understand this. Seems like we collect all assoc items in a crate already, no?

**Comment by SomeoneToIgnore** [comment]
> Seems like we collect all assoc items in a crate already, no?

Looks like we don't collect them into the `import_map` at least, my `dbg!` experiments had shown me that now we indeed start to propose assoc items for imported traits, as commented in:
https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/hir_def/src/import_map.rs#L68-L70

It's a bit weird considering that we're able to colelct all importable traits and structs into the same `fst` but not the assoc items though.
I'm not 100% fluent with the code yet and plan to investigate it later in detail, was just hoping that somebody knows what's going on there.

File: crates/completion/src/completions/unqualified_path.rs (hunk 1 above)
Comment (SomeoneToIgnore) [reply crates/completion/src/completions/unqualified_path.rs:153]: > It's even time to refactor it to CommandObject pattern, a-la FindUsage.

IMO if we still plan to move local symbol search into `import_map`, we might just wait for that: with a single place to query things, whole `imports_locator` is not needed, including the method mentioned.

Comment (SomeoneToIgnore) [reply crates/completion/src/completions/unqualified_path.rs]: That turned out a bit uglier than exlected due to an existential type returned from the method, but done.

**Comment by SomeoneToIgnore** [comment]
Actually, never mind me, I've rechecked and it looks like we're properly fine with storing and querying all traits already.
I could notice any notable change in `fuzzy_import` span timings with all those items then, interesting.

One step less towards the completions then?

File: crates/completion/src/completions/unqualified_path.rs [hunk 4]
```diff
@@ -124,8 +124,8 @@ fn complete_enum_variants(acc: &mut Completions, ctx: &CompletionContext, ty: &T
 // Note that having this flag set to `true` does not guarantee that the feature is enabled: your client needs to have the corredponding
 // capability enabled.
 fn fuzzy_completion(acc: &mut Completions, ctx: &CompletionContext) -> Option<()> {
-    let _p = profile::span("fuzzy_completion");
     let potential_import_name = ctx.token.to_string();
+    let _p = profile::span("fuzzy_completion").detail(|| potential_import_name.clone());
```
Comment (matklad) [inline crates/completion/src/completions/unqualified_path.rs:128]: :+1:

File: crates/hir_def/src/import_map.rs [hunk 5]
```diff
@@ -185,6 +186,7 @@ impl ImportMap {
         is_type_in_ns: bool,
         original_import_info: &ImportInfo,
     ) {
+        mark::hit!(type_aliases_ignored);
```
Comment (matklad) [inline crates/hir_def/src/import_map.rs:189]: let's move this to `continue` arm?

**Review by matklad — COMMENTED** [review]
r=me

**Comment by SomeoneToIgnore** [comment]
bors r=matklad

**Pull request merged**